### PR TITLE
Return nil metadata values when data is missing

### LINKED ISSUE #206: Something's awry with Supermarket 2.5.2 artifact info
When the options hash is `{channel: :stable, product_name: 'supermarket', product_version: '2.5.2'}`, we get our array of artifacts. When `include_metadata: true` is added, the gem blows up.

I don't know if maybe there's something special about that release? @nellshamrell's [commit for that version](https://github.com/chef/supermarket/commit/3e49c83c6fc4473da99bb8c87097deed976ace11) may be significant, but I don't know Omnibus so I wouldn't recognize it if it was.

This came up while chasing some errors in the downloads site logs and didn't find any other instances of the problem. Just this version and product combination.

Examples:
_without metadata_
```
*+>> Mixlib::Install.new(channel: :stable, product_name: 'supermarket', product_version: '2.5.2').artifact_info                                                                                                                        
=> [#<Mixlib::Install::ArtifactInfo:0x00000004ca3e98 @architecture="x86_64", @license=nil, @license_content=nil, @md5="15c8af5577e7186afabb14e2799de034", @platform="ubuntu", @platform_version="12.04", @product_description="Supermarket", @product_name="supermarket", @sha1="dfd5ac147e17db42092bba0fc134b29da63df222", @sha256="45631cd555c8eacf9eac0970ddd1508c5e762f77bbd49248fdfbc9e16c309d5b", @software_dependencies=nil, @url="https://packages.chef.io/files/stable/
supermarket/2.5.2/ubuntu/12.04/supermarket_2.5.2-1_amd64.deb", @version="2.5.2">, #<Mixlib::Install::ArtifactInfo:0x00000004ca3998 @architecture="x86_64", @license=nil, @license_content=nil, @md5="cd85b5d4e9006ddcc8f7325a90796ac7", @platform="el", @platform_version="5", @product_description="Supermarket", @product_name="supermarket", @sha1="e85e4c6ed47bb9a17a0b1be991caee5de150bc18", @sha256="9ef0a68dcf84474691beeff0f46315b4b992c680366ab38d5a2ae94c7423af5b", @s
oftware_dependencies=nil, @url="http://packages.chef.io/files/stable/supermarket/2.5.2/el/5/supermarket-2.5.2-1.el5.x86_64.rpm", @version="2.5.2">, #<Mixlib::Install::ArtifactInfo:0x00000004ca3498 @architecture="x86_64", @license=nil, @license_content=nil, @md5="e63c6aeb3380e1eb4ef70f68e59b0faa", @platform="el", @platform_version="6", @product_description="Supermarket", @product_name="supermarket", @sha1="0fb146868dc61399388f68304b0c35a240730fa6", @sha256="b3e
f0037aa64b2df5e5e4cb7083a7f3f3e8ea152a88e4c726c3fcaff18d3ded2", @software_dependencies=nil, @url="https://packages.chef.io/files/stable/supermarket/2.5.2/el/6/supermarket-2.5.2-1.el6.x86_64.rpm", @version="2.5.2">, #<Mixlib::Install::ArtifactInfo:0x00000004ca2f98 @architecture="x86_64", @license=nil, @license_content=nil, @md5="ba03d394a9d1ca9afb290e9b743e057a", @platform="el", @platform_version="7", @product_description="Supermarket", @product_name="supermark
et", @sha1="9e8c5a3ca1a2a51c605770abccafcb86c567f906", @sha256="c6abef85d8673037b6b5f19f56b49090d2ccbcb740b2115af87b69779eeca2fc", @software_dependencies=nil, @url="https://packages.chef.io/files/stable/supermarket/2.5.2/el/7/supermarket-2.5.2-1.el7.x86_64.rpm", @version="2.5.2">, #<Mixlib::Install::ArtifactInfo:0x00000004ca2a48 @architecture="x86_64", @license=nil, @license_content=nil, @md5="15c8af5577e7186afabb14e2799de034", @platform="ubuntu", @platform_ve
rsion="10.04", @product_description="Supermarket", @product_name="supermarket", @sha1="dfd5ac147e17db42092bba0fc134b29da63df222", @sha256="45631cd555c8eacf9eac0970ddd1508c5e762f77bbd49248fdfbc9e16c309d5b", @software_dependencies=nil, @url="https://packages.chef.io/files/stable/supermarket/2.5.2/ubuntu/10.04/supermarket_2.5.2-1_amd64.deb", @version="2.5.2">, #<Mixlib::Install::ArtifactInfo:0x00000004ca2548 @architecture="x86_64", @license=nil, @license_content=
nil, @md5="15c8af5577e7186afabb14e2799de034", @platform="ubuntu", @platform_version="14.04", @product_description="Supermarket", @product_name="supermarket", @sha1="dfd5ac147e17db42092bba0fc134b29da63df222", @sha256="45631cd555c8eacf9eac0970ddd1508c5e762f77bbd49248fdfbc9e16c309d5b", @software_dependencies=nil, @url="https://packages.chef.io/files/stable/supermarket/2.5.2/ubuntu/14.04/supermarket_2.5.2-1_amd64.deb", @version="2.5.2">]
```
_with metadata_
```
*+>> Mixlib::Install.new(channel: :stable, product_name: 'supermarket', product_version: '2.5.2', include_metadata: true).artifact_info                                                                                                
NoMethodError: undefined method `[]' for nil:NilClass                                                                                                                                                                                  
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/mixlib-install-3.2.1/lib/mixlib/install/backend/package_router.rb:190:in `create_artifact'                                                                      
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/mixlib-install-3.2.1/lib/mixlib/install/backend/package_router.rb:144:in `block in artifacts_for_version'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/mixlib-install-3.2.1/lib/mixlib/install/backend/package_router.rb:144:in `map'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/mixlib-install-3.2.1/lib/mixlib/install/backend/package_router.rb:144:in `artifacts_for_version'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/mixlib-install-3.2.1/lib/mixlib/install/backend/package_router.rb:43:in `available_artifacts'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/mixlib-install-3.2.1/lib/mixlib/install/backend/base.rb:62:in `info'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/mixlib-install-3.2.1/lib/mixlib/install/backend.rb:29:in `info'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/mixlib-install-3.2.1/lib/mixlib/install.rb:46:in `artifact_info'
        from (irb):9
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/hanami-1.0.0/lib/hanami/commands/console.rb:61:in `start'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/hanami-1.0.0/lib/hanami/cli.rb:89:in `console'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/thor-0.19.4/lib/thor/command.rb:27:in `run'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/thor-0.19.4/lib/thor/invocation.rb:126:in `invoke_command'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/thor-0.19.4/lib/thor.rb:369:in `dispatch'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/thor-0.19.4/lib/thor/base.rb:444:in `start'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/gems/hanami-1.0.0/bin/hanami:5:in `<top (required)>'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/bin/hanami:22:in `load'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/bin/hanami:22:in `<main>'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/bin/ruby_executable_hooks:15:in `eval'
        from /home/trevor/.rvm/gems/ruby-2.3.3@chef-web-downloads/bin/ruby_executable_hooks:15:in `<main>'
```

### PR DESCRIPTION
**Pull request by wrightp**
Signed-off-by: Patrick Wright <patrick@chef.io>

Fixes #206 

@chef/engineering-services @TrevorBramble

### CODE REVIEW

**Comment by wrightp** [comment]
Mixlib Install should follow suite with chefdk and chef ruby version requirements. 2.3.0 and greater

**Comment by wrightp** [comment]
This change will result in a major version bump

**Comment by wrightp** [comment]
RE: versioning https://github.com/chef/chef-rfc/blob/master/rfc034-ruby-193-eol.md
tl;dr minor bump

**Comment by wrightp** [comment]
Ok, turns out the ruby 2.3 change change will break chef client updater. For now, I can just make the change 2.2 compliant as I figure out how to make this work for everyone.

**Comment by TrevorBramble** [comment]
@wrightp anything i can do to get this into a release?
